### PR TITLE
stats: fix histogram boundaries overflow error

### LINKED ISSUE #7868: Error occurred when read table stats for table, The error message is [types:1690]constant -9223372036854775808 overflows int
## Bug Report

1. What did you do?
    update tidb cluster from 2.0.6 to 2.1.rc3

2. What did you expect to see?
    query should be very fast.

3. What did you see instead?

   query on some table is very slow, the error message is

```go
2018/10/11 09:33:54.919 handle.go:152: [debug] Error occurred when read table stats for table t1_domain_cl. The error message is [types:1690]constant -9223372036854775808 overflows int
github.com/pingcap/tidb/vendor/github.com/pkg/errors.AddStack
	/go/src/github.com/pingcap/tidb/vendor/github.com/pkg/errors/errors.go:173
github.com/pingcap/tidb/terror.(*Error).GenWithStack
	/go/src/github.com/pingcap/tidb/terror/terror.go:224
github.com/pingcap/tidb/types.overflow
	/go/src/github.com/pingcap/tidb/types/etc.go:213
github.com/pingcap/tidb/types.ConvertIntToInt
	/go/src/github.com/pingcap/tidb/types/convert.go:89
github.com/pingcap/tidb/types.(*Datum).toSignedInteger
	/go/src/github.com/pingcap/tidb/types/datum.go:1403
github.com/pingcap/tidb/types.(*Datum).convertToInt
	/go/src/github.com/pingcap/tidb/types/datum.go:845
github.com/pingcap/tidb/types.(*Datum).ConvertTo
	/go/src/github.com/pingcap/tidb/types/datum.go:672
github.com/pingcap/tidb/statistics.(*Handle).histogramFromStorage
	/go/src/github.com/pingcap/tidb/statistics/histogram.go:309
github.com/pingcap/tidb/statistics.(*Handle).columnStatsFromStorage
	/go/src/github.com/pingcap/tidb/statistics/table.go:195
github.com/pingcap/tidb/statistics.(*Handle).tableStatsFromStorage
	/go/src/github.com/pingcap/tidb/statistics/table.go:265
github.com/pingcap/tidb/statistics.(*Handle).Update
	/go/src/github.com/pingcap/tidb/statistics/handle.go:149
github.com/pingcap/tidb/domain.(*Domain).updateStatsWorker
	/go/src/github.com/pingcap/tidb/domain/domain.go:738
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:2361.
```

I have tried `analyze table t1_domain_cl` manually, it takes 2 hours to finish. but the error persists.

stats data in stats_buckets
```
TiDB-Server [mysql]> select * from stats_buckets where table_id = 511 limit 10;
+----------+----------+---------+-----------+-------+---------+-------------+----------------------+
| table_id | is_index | hist_id | bucket_id | count | repeats | upper_bound | lower_bound          |
+----------+----------+---------+-----------+-------+---------+-------------+----------------------+
|      511 |        0 |       1 |         0 |    32 |       0 | 13976888    | -9223372036854775808 |
|      511 |        0 |       1 |         1 |    33 |       0 | 22171982    | 13976889             |
|      511 |        0 |       1 |         2 |    35 |       0 | 31427171    | 22171983             |
|      511 |        0 |       1 |         3 |    34 |       0 | 42153316    | 31427172             |
|      511 |        0 |       1 |         4 |    34 |       0 | 53819134    | 42153317             |
|      511 |        0 |       1 |         5 |    33 |       0 | 62127734    | 53819135             |
|      511 |        0 |       1 |         6 |    34 |       0 | 71515151    | 62127735             |
|      511 |        0 |       1 |         7 |    33 |       0 | 81125612    | 71515152             |
|      511 |        0 |       1 |         8 |    32 |       0 | 96038585    | 81125613             |
|      511 |        0 |       1 |         9 |    35 |       0 | 107480738   | 96038586             |
+----------+----------+---------+-----------+-------+---------+-------------+----------------------+
10 rows in set (0.01 sec)
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

```
Release Version: v2.1.0-rc.3-1-g8823f12
Git Commit Hash: 8823f12d324b0d5cc4b0750df5482d0eba521d2e
Git Branch: release-2.1
UTC Build Time: 2018-10-09 07:19:27
GoVersion: go version go1.10.4 linux/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by lamxTyler**
### What problem does this PR solve? 

When loading histogram from the store, it will raise an error like `constant -9223372036854775808 overflows int`.
Fix #7868 

### What is changed and how it works?

The out of range value comes from the feedback. We should adjust their values according to the column type info.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - If the error already happens, we need to first analyze the table.

Related changes

 - Need to cherry-pick to the release branch

PTAL @zz-jason @winoros @eurekaka

### CODE REVIEW

File: statistics/feedback.go [hunk 1]
```diff
@@ -528,7 +587,12 @@ func splitBuckets(h *Histogram, feedback *QueryFeedback) ([]bucket, []bool, int6
 func UpdateHistogram(h *Histogram, feedback *QueryFeedback) *Histogram {
 	buckets, isNewBuckets, totalCount := splitBuckets(h, feedback)
 	buckets = mergeBuckets(buckets, isNewBuckets, float64(totalCount))
-	return buildNewHistogram(h, buckets)
+	hist := buildNewHistogram(h, buckets)
+	// Update the NDV of primary key column.
```
Comment (eurekaka) [inline statistics/feedback.go]: Why do we move NDV update to here?

**Comment by lamxTyler** [comment]
@eurekaka Yes, but maybe that is another problem: for example, when we have a primary key of the handle, the query range of `a < 1` will be transformed to (minInt64, 1), no matter the type of primary key, so feedback will get the lower bound as minInt64.

File: statistics/feedback.go [hunk 2]
```diff
@@ -243,20 +243,79 @@ type BucketFeedback struct {
...
 // buildBucketFeedback build the feedback for each bucket from the histogram feedback.
 func buildBucketFeedback(h *Histogram, feedback *QueryFeedback) (map[int]*BucketFeedback, int) {
 	bktID2FB := make(map[int]*BucketFeedback)
+	if len(feedback.feedback) == 0 {
+		return bktID2FB, 0
+	}
 	total := 0
-	for _, ran := range feedback.feedback {
-		idx, _ := h.Bounds.LowerBound(0, ran.lower)
+	sc := &stmtctx.StatementContext{TimeZone: time.UTC}
```
[the remark this answers is not in the document]
Comment (lamxTyler) [reply statistics/feedback.go]: Because the timeZone does not matter here, so we just do not leave it as nil.

File: statistics/feedback.go (hunk 1 above)
Comment (lamxTyler) [reply statistics/feedback.go]: - The origin place is not correct, it is before the update of the histogram.

- Here is more natural, if we update the histogram, we'd better update all the info here.

**Pull request merged**